### PR TITLE
Upgrade to support Kubernetes 1.6.x and change to Heketi-gluster fileserver

### PR DESCRIPTION
**Pull request by andersla**
## Change content and motivation
Upgrade to support Kubernetes 1.6.x and change to Heketi-gluster fileserver
Including:
- Add RBAC workaround
- Upgrade Weave and Traefik playbooks
- Add heketi-gluster playbooks
- Remove previous gluster playbooks
- Update terraform scripts (and template files) to include "glusternode"

## GitHub cross-links 

**Fixes:** 

**Docs**:

### CODE REVIEW

**Comment by mcapuccini** [comment]
@andersla this branch has some conflicts to be resolved. Could you use the stable image that @carmat88 prepared?

File: .travis.yml [hunk 1]
```diff
@@ -108,8 +111,10 @@ before_script:
 script:
   # Deploy
   - travis_retry ../bin/kn apply "$HOST_CLOUD"
+  - ../bin/kn ansible-playbook playbooks/install-heketi-gluster.yml
   # Test
   - ../bin/kn ansible-playbook playbooks/infra-test.yml
+  - ../bin/kn ansible-playbook playbooks/infra-test-more.yml
```
Comment (mcapuccini) [inline .travis.yml]: Please make a single infra-test.yml

File: aws/main.tf [hunk 2]
```diff
@@ -231,6 +248,35 @@ module "edge" {
...
+  security_group_ids = "${concat(module.security_group.id, var.additional_sec_group_ids)}"
+
+  # Disk settings
+  disk_size       = "${var.glusternode_disk_size}"
+  extra_disk_size = "${var.glusternode_extra_disk_size}"
+
+  # Bootstrap settings
+  bootstrap_file = "bootstrap/node.sh"
+  kubeadm_token  = "${var.kubeadm_token}"
+  node_labels    = ["storagenode=glusterfs"]
```
Comment (mcapuccini) [inline aws/main.tf]: why a new label? Is there any problem with calling it `role=glusternode`?

File: bootstrap/master.sh [hunk 3]
```diff
@@ -25,13 +25,26 @@ fi
 systemctl daemon-reload
 systemctl restart kubelet
 
-echo "Inititializing the master..."
+# execute modprobe on node - workaround for heketi gluster
```
Comment (mcapuccini) [inline bootstrap/master.sh]: Could you explain it. It's enough to have it here as a comment, just for reference.

File: bootstrap/master.sh [hunk 4]
```diff
@@ -25,13 +25,26 @@ fi
...
     # shellcheck disable=SC2154
-    kubeadm init --token "${kubeadm_token}" --use-kubernetes-version=v1.5.2 --api-advertise-address="$API_ADVERTISE_ADDRESSES"
+    kubeadm init --token "${kubeadm_token}" --kubernetes-version=v1.6.4 --api-advertise-address="$API_ADVERTISE_ADDRESSES"
 else
     # shellcheck disable=SC2154
-    kubeadm init --token "${kubeadm_token}" --use-kubernetes-version=v1.5.2
+    kubeadm init --token "${kubeadm_token}" --kubernetes-version=v1.6.4
 fi
+
+# Copy kubernetes configuration created by kubeadm (admin.conf to .kube/config)
```
Comment (mcapuccini) [inline bootstrap/master.sh]: Kubernetes*

File: gce/main.tf [hunk 5]
```diff
@@ -175,6 +192,34 @@ module "edge" {
...
+
+  # Disk settings
+  disk_size       = "${var.glusternode_disk_size}"
+  extra_disk_size = "${var.glusternode_extra_disk_size}"
+
+  # Bootstrap settings
+  bootstrap_file = "bootstrap/node.sh"
+  kubeadm_token  = "${var.kubeadm_token}"
+  node_labels    = ["storagenode=glusterfs"]
+  node_taints    = [""]                                       # dedicated=fileserver:NoSchedule
```
Comment (mcapuccini) [inline gce/main.tf]: Shouldn't be tainted?

File: playbooks/infra-test-more.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+- hosts: master
```
Comment (mcapuccini) [inline playbooks/infra-test-more.yml]: please make a single `infra-test.yml`

File: playbooks/install-heketi-gluster-pvc.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+- hosts: master
```
Comment (mcapuccini) [inline playbooks/install-heketi-gluster-pvc.yml]: please rename it `create-gluster-pvc.yml`

File: playbooks/install-heketi-gluster.yml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+- hosts: master
```
Comment (mcapuccini) [inline playbooks/install-heketi-gluster.yml]: this should be part of `install-core.yml`

File: gce/main.tf (hunk 5 above)
Comment (mcapuccini) [inline gce/main.tf]: please remove the comment if not

File: playbooks/roles/common/tasks/main.yml [hunk 9]
```diff
@@ -1 +1,3 @@
 - include: wait-nodes.yml
+- include: workaround-RBAC.yml
+- { include: untaint_master.yml, when: nodes_count == 1 }
```
Comment (mcapuccini) [inline playbooks/roles/common/tasks/main.yml]: this doesn't belong to this PR I guess it's for single node deployment right?

File: playbooks/roles/common/tasks/untaint_master.yml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+- name: "untaint master if it is the only node"
+  command: >
+    kubectl taint nodes
+    --all node-role.kubernetes.io/master-
```
Comment (mcapuccini) [inline playbooks/roles/common/tasks/untaint_master.yml]: doesn't belong to this PR

File: playbooks/roles/heketi-gluster-server/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,40 @@
+---
+- name: create gluster kubenow yaml directory
```
Comment (mcapuccini) [inline playbooks/roles/heketi-gluster-server/tasks/main.yml]: KubeNow*

File: playbooks/roles/heketi-gluster-server/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  copy:
+    src: glusterfs-daemonset.yml
+    dest: "~/.kubernetes-yaml/glusterfs/glusterfs-daemonset.yml"
+
+- name: create gluster DaemonSet
+  command: >
+    kubectl apply -f
+    $HOME/.kubernetes-yaml/glusterfs/glusterfs-daemonset.yml
+
+# the plays below is not really needed - just for verifying everything is OK
```
Comment (mcapuccini) [inline playbooks/roles/heketi-gluster-server/tasks/main.yml]: please remove this comment, verification is always needed 🙂

File: playbooks/roles/heketi-pvc/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (mcapuccini) [inline playbooks/roles/heketi-pvc/tasks/main.yml]: we don't want to create PVC by default, please remove this

File: playbooks/roles/heketi-storage/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,68 @@
+- name: retrieve gluster-pod IPs
```
Comment (mcapuccini) [inline playbooks/roles/heketi-storage/tasks/main.yml]: we have too many different directories hekety, hekety-storage, gluster-server-heketi, please make a single one

File: playbooks/roles/weave-network/files/weave-daemonset-k8s-v1.9.5.yaml [hunk 15]
```diff
@@ -0,0 +1,118 @@
+kind: ClusterRole
```
Comment (mcapuccini) [inline playbooks/roles/weave-network/files/weave-daemonset-k8s-v1.9.5.yaml]: Can't we find it versioned on the internet? Or do they only provide unversioned Weave?

File: templates/terraform.tfvars.aws-template [hunk 16]
```diff
@@ -23,6 +23,12 @@ node_disk_size = "10" # Size in GB
 # edge_instance_type = "t2.medium"
 # edge_disk_size = "10" # Size in GB
 
+# Gluster configuration
+# glusternode_count = "2"
+# glusternode_instance_type = "t2.medium"
+# glusternode_disk_size = "10" # Size in GB
```
Comment (mcapuccini) [inline templates/terraform.tfvars.aws-template]: please increase to 20 like in gce

File: aws/main.tf (hunk 2 above)
Comment (andersla) [inline aws/main.tf]: That's how they do it in heketi-project

File: bootstrap/master.sh (hunk 3 above)
Comment (andersla) [inline bootstrap/master.sh]: Somehow this module doesn't get loaded into system by default. In Heketi project they do this manually in a Ansible playbook. module is apparently part of LMV and needed for heketi

Comment (mcapuccini) [inline bootstrap/master.sh]: 👍

File: playbooks/roles/heketi-gluster-server/tasks/main.yml (hunk 12 above)
Comment (andersla) [inline playbooks/roles/heketi-gluster-server/tasks/main.yml]: What I mean is that sometimes we turn the deployment into more serial than what is needed with kubernetes, we are waiting for everything to come up before moving on, when more tasks could be parallelllized.

File: playbooks/roles/weave-network/files/weave-daemonset-k8s-v1.9.5.yaml (hunk 15 above)
Comment (andersla) [inline playbooks/roles/weave-network/files/weave-daemonset-k8s-v1.9.5.yaml]: Looking around again - I found versions of the daemonset yaml on github release page:-)

File: templates/terraform.tfvars.aws-template (hunk 16 above)
Comment (andersla) [inline templates/terraform.tfvars.aws-template]: And if we are doing gluster nodes default I will lower count to 1

File: playbooks/roles/common/tasks/main.yml (hunk 9 above)
Comment (andersla) [inline playbooks/roles/common/tasks/main.yml]: You forgot to write "but it doesn't matter" so I moved it to another pull request

**Comment by mcapuccini** [comment]
@andersla there is a `install-heketi-gluster.yml`, maybe it should be removed now?

**Comment by mcapuccini** [comment]
It's alive 🙂
